src/command.rs

```diff
@@ -0,0 +1,306 @@
+    );
+
+    const DIR_ERR: &str = "Failed to read current directory name";
+
+    let name = match name {
+        Some(name) => name,
+        None => std::env::current_dir()?
+            .file_name()
+            .wrap_err(DIR_ERR)?
+            .to_str()
+            .wrap_err(DIR_ERR)?
+            .parse()?,
+    };
+
+    const VERSION: Version = Version::new(0, 1, 0);
+
+    let manifest = Manifest {
+        package: PackageManifest {
+            r#type,
+            name,
+            version: VERSION,
+            description: None,
+        },
+        dependencies: vec![],
+    };
+
+    manifest.write().await?;
+
+    PackageStore::create().await
+}
+
+/// Adds a dependency to this project
+pub async fn add(dependency: String) -> eyre::Result<()> {
+    let lower_kebab = |c: char| (c.is_lowercase() && c.is_ascii_alphabetic()) || c == '-';
+
+    let (repository, dependency) = dependency
+        .trim()
+        .split_once('/')
+        .wrap_err("Invalid dependency specification")?;
+
+    ensure!(
+        repository.chars().all(lower_kebab),
+        "Repositories must be in lower kebab case"
+    );
+
+    let (package, version) = dependency
+        .split_once('@')
+        .wrap_err_with(|| format!("Invalid dependency specification: {dependency}"))?;
+
+    let package = package
+        .parse::<PackageName>()
+        .wrap_err_with(|| format!("Invalid package name supplied: {package}"))?;
+
+    let version = version
+        .parse::<VersionReq>()
+        .wrap_err_with(|| format!("Invalid version requirement supplied: {package}"))?;
+
+    let mut manifest = Manifest::read().await?;
+
+    manifest
+        .dependencies
+        .push(Dependency::new(repository.to_owned(), package, version));
+
+    manifest.write().await
+}
+
+/// Removes a dependency from this project
+pub async fn remove(package: PackageName) -> eyre::Result<()> {
+    let mut manifest = Manifest::read().await?;
+
+    let dependency = manifest
+        .dependencies
+        .iter()
+        .find(|d| d.package == package)
+        .wrap_err(eyre::eyre!(
+            "Unable to remove unknown dependency {package:?}"
+        ))?
+        .to_owned();
+
+    manifest.dependencies.retain(|d| *d != dependency);
+
+    PackageStore::uninstall(&dependency.package).await.ok();
+
+    manifest.write().await
+}
+
+/// Packages the api and writes it to the filesystem
+pub async fn package(directory: String, dry_run: bool) -> eyre::Result<()> {
+    let package = PackageStore::release()
+        .await
+        .wrap_err("Failed to create release")?;
+
+    let path = Path::new(&directory).join(format!(
+        "{}-{}.tgz",
+        package.manifest.package.name, package.manifest.package.version
+    ));
+
+    if !dry_run {
+        std::fs::write(path, package.tgz).wrap_err("failed to write package to filesystem")?;
+    }
+
+    Ok(())
+}
+
+/// Publishs the api package to the registry
+pub async fn publish(
```

Comment:
I would be highly tempted to put the version in the dependency only at publish time. It's not atypical and avoids the issue we end up having a lot where we leave repos dirty post publish.

Reply:
Can you elaborate on this? I don't follow how the repository may end up dirty post publish.

Reply:
I meant the version in the manifest. It’s typical to leave versions in manifests permanently at 0.0.0 or whatever and then when one publishes, they first rewrite the manifest, and then publish. It’d be nice if buffrs got the version out of the manifest, only adding it at publish time.

Reply:
I'm not sure this is meant to be the role of the package manager. We've been mimicking Cargo's design and it doesn't do anything special about the package version, it just reads it from the manifest. Updating the lockfile is one thing, but messing with the manifest behind your back sounds iffy.

Reply:
I think the "we use 0.0.0" as placeholder is something that is user specific / out of scope

Reply:
Out of scope in terms of what, exactly, Mara? Cargo made that design choice because it means that a crate is just a tar file - you can process a git, source, registry dependency in exactly the same way. Same thing for Python, for the same reason. But it's inconvenient for anyone using git versions, which are relatively commonly used - it means you either have a release step where you update the versions in your repo and commit that or you have an overwriting step at publish time. Avoiding needing to write down your own version in your manifest avoids that extra step.

Put another way:
- buffrs publishing as done by us internally seemingly requires overwriting of the buffrs manifest pre-publish to make sure the versions line up.
- buffrs also blows up if you have a dirty repo at publish time

How do we work around this? Doing a temporary commit of the buffrs manifest pre-push?

Reply:
I think the way it's supposed to work is that you tie publishing to releases. So as a preparatory step you add a (permanent) commit that bumps the package version, then you tag that commit with a matching (eg, GitHub) release, and CI takes care of publishing the release (or alternatively you do it locally).

I get where you are coming from though. Keeping the manifest and repository versions in sync is riddled with pitfalls and poor UX. One generally needs sophisticated scripting just to glue those together with minimal risk of mistakes, though nothing stops someone from bypassing that. Ideally versions would be unified so bumps are transactional and there's no chance of desync between repository and registry.

It's not just Cargo and PIP that work this way, btw. Every single package manager that I know has the same issue. That doesn't mean we can't take a different approach and integrate with git / version control, but since that diverges from the beaten path it'd need thoughtful planning.

Let's move this discussion to its own space, as it's another instance of something that doesn't directly tie to the Lockfile functionality.

Tracking in #68